docs/documentation/design/deducer-design.zh-CN.md

```diff
@@ -0,0 +1,232 @@
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，两次传入的参数分别是 `queue1` 和 `queue2`，分别对应于 `queue1`、`queue2` 两个特殊对象。
+
+```typescript
+function createAndConfigQueue(name: string, kv): resource.Queue {
+  const queue = new resource.Queue(name /* other configuration */);
+  // do something
+  return queue;
+}
+
+function buildQueue(queueName: string, buildClosure: Function): resource.Queue {
+  return buildClosure(queueName);
+}
+
+const queue1 = buildQueue("queue1", createAndConfigQueue);
+const queue2 = buildQueue("queue2", createAndConfigQueue);
+```
+
+**（2）隐含特殊对象**
+
+该例子中，特殊类型的构造函数返回没有赋值给变量，而是直接调用了该返回值的特殊方法`subscribe`。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，且参数为空，这次实例化过程产生了一个特殊对象；
+2. 随后调用了这个特殊对象的 `subscribe` 特殊方法，传入此次调用的参数是一个函数；
+3. Deducer 将该函数提取为闭包函数。
+
+```typescript
+new resource.Queue().subscribe(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+**（3）函数入参做特殊方法的调用方**
+
+该例子中，特殊方法 `push` 在函数 `pushOneMessage` 被调用，而调用方，则是根据函数入参确定。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，两次传入的参数分别是 `queue1` 和 `queue2`，分别对应于 `queue1`、`queue2` 两个特殊对象。
+2. `queue.push` 特殊方法被调用一次，调用方是特殊对象 `queue1` ，传入的参数是 `"Hello, Pluto!"`。
+
+```typescript
+const queue1 = new resource.Queue("queue1");
+const queue2 = new resource.Queue("queue2");
+
+function pushOneMessage(queue: resource.Queue, msg: string) {
+  queue.push(msg);
+}
+
+pushOneMessage(queue1, "Hello, Pluto!");
+```
+
+**（4）函数参数是函数返回值**
+
+该例子中，特殊类型 `resource.Queue` 的构造函数传入的参数是 `getName()` 的返回值。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，传入的参数是 `queueName`，对应 `queue` 是一个特殊对象。
+
+```typescript
+function getName(): string {
+  return "queueName";
+}
+
+const queue = new resource.Queue(getName());
+```
+
+**（5）特殊方法重命名，间接访问**
+
+该例子中，特殊对象 `queue` 的特殊方法 `subscribe` 被赋值给 subFunc 函数变量，随用调用了该函数变量。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，传入的参数是 `queue`，对应 `queue` 是一个特殊对象；
+2. 特殊方法 `subscribe` 被调用一次，调用方是特殊对象 `queue`，传入的参数是一个函数；
+3. 将该函数参数提取成闭包。
+
+```typescript
+const queue = new resource.Queue("queue");
+
+const subFunc = queue.subscribe.bind(queue);
+
+subFunc(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+**（6）动态资源对象访问，间接访问**
+
+该例子中，构建了一个 JS 对象 `queues`，该对象包含键值对，值都为特殊对象。随后，通过索引访问其中一个特殊对象。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，传入的参数分别是 `queue1`和 `queue2`，对应有两个特殊对象；
+2. 特殊方法 `subscribe` 被调用一次，调用方是 `queues["one"]` 对应的特殊对象，传入的参数是一个函数；
+3. 将该函数参数提取成闭包。
+
+```typescript
+const queues = {
+  one: new resource.Queue("queue1"),
+  two: new resource.Queue("queue2"),
+};
+
+queues["one"].subscribe(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+## 所需的原子能力
+
+从以上描述与实例可以总结，Deducer 所需的能力包括如下 4 个：
+
+1. **特殊对象的构造过程查找**：确定程序代码调用了哪些特殊类型的构造函数，进而创建了哪些特殊对象，对应示例 1、2、3、4、5、6。
+2. **特殊方法的调用过程查找**：确定程序代码中哪些特殊对象的哪些特殊方法被调用，对应示例 2、3、5、6。
+3. **值演算**：
+   - 一种情况用于在分析构造函数、特殊方法的调用时，确定传入的非函数类型参数的具体值是多少，对应示例 1、3、4、5；
+   - 另一情况用于确定特殊方法的调用方具体是哪个特殊对象，对应示例 3、6。
+4. **闭包提取**：用于在分析构造函数、特殊方法的调用时，将传入的函数类型参数提取成闭包，对应示例 2、5、6。
+
+## 原子能力的实现思路
```

Comment:
There should be no problem with the overall design and implementation ideas, and two additional points should be added

+ Grammar rules are inexhaustible, modeling with symbolic systems as much as possible.
+ Typescript and Python are both progressive type systems for analyzers. What should be done when certain symbol types cannot be obtained or when any type cannot be obtained.

Reply:
The prerequisite for Deducer to work is to be able to find special types and special methods. If there is no type information in the program code, it will directly report an error and exit. If there is some use of the any type, a warning message is given.

For the special objects, Deducer requires that their type cannot change. Special objects cannot be copied to variables of other types, and variables of other types cannot be assigned or forcibly type-casted to special objects.

If type loss occurs when performing value calculations on parameter values, and it affects the inference of values, an error is reported and the location is indicated. Otherwise, it is ignored.